dojo/unittests/tools/test_nexpose_parser.py

```diff
@@ -84,19 +84,19 @@ def test_nexpose_parser_has_many_finding(self):
         finding = findings[29]
         self.assertIn("Backup Exec Agent Browser", finding.description)
         self.assertEqual("backup-exec-agent-browser", finding.unsaved_tags[0])
-        self.assertEqual("backup-exe", finding.unsaved_endpoints[0].protocol)
+        self.assertEqual("backup-exec-agent-br", finding.unsaved_endpoints[0].protocol)
```

Comment:
@damiencarol, `20` still doesn't cover this use-case. But it is really really edge case

Reply:
WTF, realt 20??? :/
![image](https://user-images.githubusercontent.com/1694940/122794181-421fd880-d2bc-11eb-906d-96202baf3b4e.png)

Reply:
Something is wrong with this parser and I'm glad that the new code on endpoints found it.
Take a look at the data:
```xml
                           <endpoint protocol="tcp" port="6101" status="open">
					<services>
						<service name="Backup Exec Agent Browser">
							<configuration>
								<config name="sslv3">false</config>
								<config name="tlsv1_0">false</config>
								<config name="tlsv1_1">false</config>
								<config name="tlsv1_2">false</config>
							</configuration>
							<tests>
							</tests>
						</service>
					</services>
				</endpoint>
```

The protocol value should be `tcp` and not some fake value `backup-exec-agent-browser`

Reply:
Well, yes and no. What about [this case](https://github.com/DefectDojo/django-DefectDojo/blob/0a3cce0ad07b67fa4058475c964b2d8a397202c6/dojo/unittests/scans/nexpose/many_vulns.xml#L488-L490)?
```xml
<endpoint protocol="tcp" port="80" status="open">
  <services>
    <service name="HTTP">
```

Or [this](https://github.com/DefectDojo/django-DefectDojo/blob/0a3cce0ad07b67fa4058475c964b2d8a397202c6/dojo/unittests/scans/nexpose/many_vulns.xml#L145-L147)
```xml
<endpoint protocol="tcp" port="22" status="open">
  <services>
    <service name="SSH">
```

Or you can compare it with [this](https://github.com/DefectDojo/django-DefectDojo/blob/0a3cce0ad07b67fa4058475c964b2d8a397202c6/dojo/unittests/scans/nexpose/dns.xml#L50-L52)
```xml
<endpoint protocol="udp" port="53" status="open">
  <services>
      <service name="DNS">
```
I like [this](https://github.com/DefectDojo/django-DefectDojo/blob/0a3cce0ad07b67fa4058475c964b2d8a397202c6/dojo/unittests/tools/test_nexpose_parser.py#L184-L185) format: `dns://192.168.1.1#tcp` and `dns://192.168.1.1#udp`.

Reply:
We can use (in Nexpose parser) checking of existence in [`hyperlink._url.SCHEME_PORT_MAP`](https://github.com/python-hyper/hyperlink/blob/153173795621f317178e6161b2e2f6113e69aa64/src/hyperlink/_url.py#L408) like [here](https://github.com/DefectDojo/django-DefectDojo/blob/0a3cce0ad07b67fa4058475c964b2d8a397202c6/dojo/endpoint/utils.py#L7) and [here](https://github.com/DefectDojo/django-DefectDojo/blob/0a3cce0ad07b67fa4058475c964b2d8a397202c6/dojo/endpoint/utils.py#L38)
So, it is exist in `hyperlink._url.SCHEME_PORT_MAP`, use it, otherwise, use `tcp` or `udp`

Reply:
Saw your commit, it's good.

Reply:
Your comment are valid. Parsing protocol is tricky. I'm not surprised that the module `hyperlink` have internal mapping.